src/controller/artifact/processor/default.go

```diff
@@ -44,16 +55,113 @@ func (d *defaultProcessor) GetArtifactType() string {
+	artifactConfig := artifact.ExtraAttrs
+	schemaMap, ok := artifactConfig[schemaV1.XHarborAttributes]
+	if !ok {
+		return nil
+	}
+	schema := &schemaV1.Schema{}
+	mapstructure.Decode(schemaMap, schema)
+	additionTypes := make([]string, 0)
+	for _, addition := range schema.Additions {
+		additionTypes = append(additionTypes, addition.Name)
+	}
+	return additionTypes
 }
-func (d *defaultProcessor) AbstractMetadata(ctx context.Context, manifest []byte, artifact *artifact.Artifact) error {
-	// do nothing currently
-	// we can extend this function to abstract the metadata in the future if needed
+func (d *defaultProcessor) AbstractMetadata(ctx context.Context, artifact *artifact.Artifact, manifest []byte) error {
+	if artifact.ManifestMediaType != v1.MediaTypeImageManifest {
+		return nil
+	}
+	// get manifest
+	mani := &v1.Manifest{}
+	if err := json.Unmarshal(manifest, mani); err != nil {
+		log.Warningf("unsupported manifest: %v", err)
+		return nil
+	}
+	// get config layer
+	_, blob, err := d.RegCli.PullBlob(artifact.RepositoryName, mani.Config.Digest.String())
+	if err != nil {
+		log.Warningf("get config layer error: %v", err)
+		return nil
+	}
+	defer blob.Close()
+	// parse metadata from config layer
+	metadata := map[string]interface{}{}
+	if err := json.NewDecoder(blob).Decode(&metadata); err != nil {
+		log.Warningf("parse config layer error: %v", err)
+		return nil
+	}
+	// parse schema
+	schemaMap, ok := metadata[schemaV1.XHarborAttributes]
+	if !ok {
+		return nil
+	}
+	schema := &schemaV1.Schema{}
+	err = mapstructure.Decode(schemaMap, schema)
+	if err != nil {
+		log.Warningf("unsupported artifact config schema: %v", err)
+		return nil
+	}
+	// check schema version
+	if schema.SchemaVersion != schemaV1.SchemaVersionV1 {
+		log.Warningf("unsupported artifact config schema version %d", schema.SchemaVersion)
+		return nil
+	}
+	// check icon url
+	_, err = url.ParseRequestURI(schema.Icon)
+	if err != nil {
+		log.Warningf("icon url invalid %v", err)
+		schema.Icon = ""
+		m := make(map[string]interface{})
+		b, _ := json.Marshal(schema)
+		_ = json.Unmarshal(b, &m)
+		metadata[schemaV1.XHarborAttributes] = m
+	}
+	// move keys in skipKeyList
+	if schema.SkipKeyList != nil {
+		for _, denyKey := range schema.SkipKeyList {
+			delete(metadata, denyKey)
+		}
+	}
+	artifact.ExtraAttrs = metadata
+
 	return nil
 }
 func (d *defaultProcessor) AbstractAddition(ctx context.Context, artifact *artifact.Artifact, addition string) (*Addition, error) {
-	// return error directly
-	return nil, errors.New(nil).WithCode(errors.BadRequestCode).
-		WithMessage("the processor for artifact %s not found, cannot get the addition", artifact.Type)
+	schemaMap, ok := artifact.ExtraAttrs[schemaV1.XHarborAttributes]
+	if !ok {
+		return nil, errors.New(nil).WithCode(errors.BadRequestCode).
+			WithMessage("the processor for artifact %s not found, cannot get the addition", artifact.Type)
+	}
+	schema := &schemaV1.Schema{}
+	mapstructure.Decode(schemaMap, schema)
```

Comment:
handle error

Reply:
Same as ```ListAdditionTypes()```, we read artifact config from database, which these data was checked when AbstractMetadata(), so I didn't handle error here. Is it neccessary?